### ISSUE TITLE
✨ Log warning when using deprecated APIs

### ISSUE THREAD

**Issue by helderco** [dagger/dagger#4855]
### What are you trying to do?

API fields that are marked as deprecated only get codegen'ed into the docblock. That shows in reference docs, and maybe the IDE helps make it more visible as well (depending on language), but we should log a warning when these calls are being used.

### Why is this important to you?

Logging a warning makes it more visible whenever you're actually running a pipeline, no matter if you use a smart IDE or consult the reference docs during development. Removing the warning acts as encouragement to migrate the new behavior.

**Comment by TomChv** [comment]
@helderco Is it a priority? I can work on this ASAP

**Comment by helderco** [comment]
Not a priority, but may be easy to do. We no longer have any deprecated apis for now though. Python has this, only missing in Go and Node.js.

**Comment by TomChv** [comment]
@helderco I think instead of doing the job on the SDK side, we should log warning in our core API using progrock, this would be a generic and durable solution :)

WDYT @vito ?

**Comment by TomChv** [comment]
@vito @helderco Any thought on this so we can make progress?

> I think instead of doing the job on the SDK side, we should log warning in our core API using progrock, this would be a generic and durable solution :)

**Comment by helderco** [comment]
Yeah, been thinking about this because we have pending deprecations. For example:
- https://github.com/dagger/dagger/pull/7136

I added this warning:

https://github.com/dagger/dagger/blob/fc6a74297956c41801d36081d841c7410ac9651e/core/schema/container.go#L582

The intent was for the warning to show for the user when this API is used, both in the TUI and hopefully in cloud traces as well. Nothing was shown last time I ran this.

I defer to @vito to help point out what needs to be done to make this happen.

**Comment by vito** [comment]
@helderco Yeah, I think it's time to start integrating with OpenTelemetry logging properly, which I take to mean integrating `slog` with `ctx` and setting OpenTelemetry's severity levels etc. instead of mimicking stdout/stderr.

I had an idea for a "global logger" that we could use sort of like Progrock's previous logging system:

https://github.com/dagger/dagger/blob/4ec5818fab06e4ac42e4b0a74dd913e0d2c02a85/telemetry/logging.go#L40-L47

...but it's half-baked at the moment. I wanted to have the TUI and Cloud look for that special `dagger.io/global` value and use that to bubble up those logs to the top level, but that's currently not implemented. The nice thing is we can also keep the logs associated to a particular span (via `ctx`), so theoretically we can have some sort of "jump to relevant span" mechanic while still showing the logs globally.

**Comment by TomChv** [comment]
I unassigned myself from the issue since it should be done in the core engine.

**Comment by TomChv** [comment]
@helderco I'm not sure I understand how to implement that, if I just log deprecation warning in the function I'm not sure it's going to be enough, for example

```go
func (s *socketSchema) Install() {
	dagql.Fields[*core.Query]{
		dagql.Func("socket", s.socket).
			Doc("Loads a socket by its ID.").
			Deprecated("Use `loadSocketFromID` instead."),
	}.Install(s.srv)

	dagql.Fields[*core.Socket]{}.Install(s.srv)
}

type socketArgs struct {
	ID core.SocketID
}

func (s *socketSchema) socket(ctx context.Context, parent *core.Query, args socketArgs) (dagql.Instance[*core.Socket], error) {
	slog.Warn("socket is deprecated, use loadSocketFromID instead")

	return args.ID.Load(ctx, s.srv)
}
```

**Comment by helderco** [comment]
> @helderco I'm not sure I understand how to implement that, if I just log deprecation warning in the function I'm not sure it's going to be enough

Yes, that's the current state of things. The goal is to make it work. See https://github.com/dagger/dagger/issues/4855#issuecomment-2098721537.

**Comment by TomChv** [comment]
@vito Do you have any hints on where I could add this kind of logic, I see that we configure the TUI in the `dagger/main.go`

https://github.com/dagger/dagger/blob/7e99ef68b6338953096580e1a73c622e79f9bf13/cmd/dagger/main.go#L277-L312

But I do not understand how I could create a static view at the top of the logging, like a permanent span or an extra global logger like you mentionned in https://github.com/dagger/dagger/issues/4855#issuecomment-2098721537

I have an idea but I'm not sure it's the right one, I would like to create a 2nd span window, that only display messages of our choices (selected by `dagger.io/global` for example) and these message would not disapears + will stay at the top of the execution.

How could I achieve such result?

**Comment by vito** [comment]
@TomChv

> I have an idea but I'm not sure it's the right one, I would like to create a 2nd span window, that only display messages of our choices (selected by `dagger.io/global` for example) and these message would not disapears + will stay at the top of the execution.

Yeah this sounds good. I actually just got done consolidating various forms of 'log panels' (we had 3!) to save screen real estate, so I'm hesitant to add another, but this feels warranted because a) it should be for "unusual" logs that we want to keep onscreen in any context, and b) we also want to let you jump to the log's originating span, which is a special mechanic not applied to other logs since they're already in the context of a span.

So as long as the TUI overhead is small, sgtm.

As far as implementation, would prefer to wait until #7671 is merged since it changes a ton of things. I figure we can intercept the log exporter and build up a list of log records that have the `dagger.io/logs.global` attribute, and then render them in `Render()` and `finalRender()`. That's probably a good enough start. The cherry on top would be the ability to somehow jump to their originating span, but that increases the difficulty a bit.
